src/lib/InstantSearch.ts

```diff
@@ -480,6 +482,20 @@ See ${createDocumentationLink({
       });
     });
 
+    // if the instance is being re-started,
+    if (this.disposed) {
+      this.disposed = false;
+      this.middleware = this.middleware.map(({ creator }) => ({
```

Comment:
it's similar to what we have in `use()`: https://github.com/algolia/instantsearch.js/blob/25f41b2b3dcc71eb5b1c5598d98808d5cf2b023f/src/lib/InstantSearch.ts#L278:L302

but `use()` can be called either before or after `start()`.

Reply:
maybe .use could be called somehow?

Reply:
yes,
```js
      this.middleware = this.middleware.map(({ creator }) => ({
        creator,
        instance: {
          subscribe: noop,
          unsubscribe: noop,
          onStateChange: noop,
          ...creator({ instantSearchInstance: this }),
        },
      }));
```

can become

```js
const middlewaresToUse = this.middleware.map(({ creator }) => creator);
this.middleware = [];
this.use(...middlewares)
```

I didn't use `use()` function in this implementation because `use()` includes irrelevant code like calling `subscribe()` if the instantSearchInstance is already started. The second version looks short, but I feel like it creates unnecessary link to the function.

What do you think? The current version is too verbose?

Reply:
I actually don't really understand why the middleware needs to be recreated, should the state not be set at `subscribe` (instead at function body now for routing middleware) so that it doesn't become stateful on creation, but on subscribe

Reply:
Ah okay, about that part, initially I tried to make `subscribe()` be the place where it sets the state from the URL (Fix 1 or Fix 2 in the PR description). And it didn't work, and this was the final try. But reading your comment, I've become to not like making middlewares stateful on creation. Fix 1 and 2 are breaking the current behaviors. Do you have any other idea? If not, I should probably close this for now, and revisit it later for next major(, although as we've recently said, it's hard to prioritize these small/invisible tasks when working on major.)

Reply:
why did moving the code to subscribe not work? I feel like that's the correct solution

Reply:
> Subscribe is called on start & restart right?

yes. And moving `_initialUiState = ....` to `subscribe()` didn't work because 

Inside `search.start()`,

```js
mainIndex.init({
  // ...
  uiState: this._initialUiState,
});

(each middleware).subscribe();
```

It's already too late.